src/video_core/page_manager.cpp

```diff
@@ -27,207 +21,49 @@ namespace VideoCore {
-    }
-
-    static LONG WINAPI GuestFaultSignalHandler(EXCEPTION_POINTERS* pExp) noexcept {
-        const u32 ec = pExp->ExceptionRecord->ExceptionCode;
-        if (ec == EXCEPTION_ACCESS_VIOLATION) {
-            const auto info = pExp->ExceptionRecord->ExceptionInformation;
-            if (info[0] == 1) { // Write violation
-                const VAddr addr_aligned = Common::AlignDown(info[1], PAGESIZE);
-                rasterizer->InvalidateMemory(addr_aligned, PAGESIZE);
-                return EXCEPTION_CONTINUE_EXECUTION;
-            } /* else {
-                UNREACHABLE();
-            }*/
-        }
-        return EXCEPTION_CONTINUE_SEARCH; // pass further
-    }
-
-    inline static Vulkan::Rasterizer* rasterizer;
-    void* veh_handle{};
-};
-#elif ENABLE_USERFAULTFD
-struct PageManager::Impl {
-    Impl(Vulkan::Rasterizer* rasterizer_) : rasterizer{rasterizer_} {
-        uffd = syscall(__NR_userfaultfd, O_CLOEXEC | O_NONBLOCK);
-        ASSERT_MSG(uffd != -1, "{}", Common::GetLastErrorMsg());
-
-        // Request uffdio features from kernel.
-        uffdio_api api;
-        api.api = UFFD_API;
-        api.features = UFFD_FEATURE_THREAD_ID;
-        const int ret = ioctl(uffd, UFFDIO_API, &api);
-        ASSERT(ret == 0 && api.api == UFFD_API);
-
-        // Create uffd handler thread
-        ufd_thread = std::jthread([&](std::stop_token token) { UffdHandler(token); });
+        // Should be called first.
+        constexpr auto priority = std::numeric_limits<u32>::min();
+        Core::Signals::Instance()->RegisterAccessViolationHandler(GuestFaultSignalHandler,
+                                                                  priority);
     }
 
     void OnMap(VAddr address, size_t size) {
-        uffdio_register reg;
-        reg.range.start = address;
-        reg.range.len = size;
-        reg.mode = UFFDIO_REGISTER_MODE_WP;
-        const int ret = ioctl(uffd, UFFDIO_REGISTER, &reg);
-        ASSERT_MSG(ret != -1, "Uffdio register failed");
+        owned_ranges += boost::icl::interval<VAddr>::right_open(address, address + size);
     }
 
     void OnUnmap(VAddr address, size_t size) {
-        uffdio_range range;
-        range.start = address;
-        range.len = size;
-        const int ret = ioctl(uffd, UFFDIO_UNREGISTER, &range);
-        ASSERT_MSG(ret != -1, "Uffdio unregister failed");
+        owned_ranges -= boost::icl::interval<VAddr>::right_open(address, address + size);
     }
 
     void Protect(VAddr address, size_t size, bool allow_write) {
-        uffdio_writeprotect wp;
-        wp.range.start = address;
-        wp.range.len = size;
-        wp.mode = allow_write ? 0 : UFFDIO_WRITEPROTECT_MODE_WP;
-        const int ret = ioctl(uffd, UFFDIO_WRITEPROTECT, &wp);
-        ASSERT_MSG(ret != -1, "Uffdio writeprotect failed with error: {}",
-                   Common::GetLastErrorMsg());
-    }
-
-    void UffdHandler(std::stop_token token) {
```

Comment:
Userfaultfd implementation seems to have been removed. Since its still useful should probably be re-added

Reply:
It doesn't really work for the new central signal handling component, but I've re-added the `PageManager` implementation.